torch/csrc/autograd/profiler.cpp

```diff
@@ -13,135 +13,260 @@
+                  s << ", ";
+                }
+              }
+              s << "]";
+            } else {
+              s << "[]";
+            }
+            if (i < shapes.size() - 1) {
+              s << ", ";
             }
-            s << "]";
           }
-          else
-            s << "[]";
-          if(i < shapes.size() - 1)
-            s << ", ";
+          s << "]";
         }
-        s << "]";
+        cuda_stubs->nvtxRangePushA(s.str().c_str());
+      } else {
+        cuda_stubs->nvtxRangePushA(name.str());
       }
-      cuda_stubs->nvtxRangePushA(s.str().c_str());
     } else {
-      cuda_stubs->nvtxRangePushA(name.str());
+      std::lock_guard<std::mutex> guard(mutex_);
+      auto& list = getEventList();
+      list.record(
+          EventKind::PushRange,
+          name,
+          RecordFunction::currentThreadId(),
+          config_.state == ProfilerState::CUDA,
+          std::move(shapes));
     }
-  } else {
-    getEventList().record(
-        EventKind::PushRange,
-        name,
-        thread_id,
-        state == ProfilerState::CUDA,
-        std::move(shapes));
   }
-}
 
-void popRange() {
-  if (state == ProfilerState::Disabled) {
-    return;
-  }
-  if (state == ProfilerState::NVTX) {
-    cuda_stubs->nvtxRangePop();
-  } else {
-    getEventList().record(
-        EventKind::PopRange,
-        StringView(""),
-        thread_id,
-        state == ProfilerState::CUDA);
+  void popRange(uint64_t orig_thread_id) {
+    if (config_.state == ProfilerState::Disabled) {
+      return;
+    }
+    if (config_.state == ProfilerState::NVTX) {
+      cuda_stubs->nvtxRangePop();
+    } else {
+      std::lock_guard<std::mutex> guard(mutex_);
+      auto& list = getEventList(orig_thread_id);
+      list.record(
+          EventKind::PopRange,
+          StringView(""),
+          orig_thread_id,
+          config_.state == ProfilerState::CUDA);
+    }
   }
-}
 
-void enableProfiler(ProfilerConfig config) {
-  ProfilerState new_state = config.state;
-  AT_ASSERT(new_state != ProfilerState::Disabled);
-  if (new_state == ProfilerState::NVTX && !cuda_stubs->enabled())
-    throw std::runtime_error("Can't use NVTX profiler - PyTorch was compiled without CUDA");
-  if (state != ProfilerState::Disabled && new_state != state) {
-    throw std::runtime_error("can't change kind of profiling (e.g. NVTX to CPU) while profiler is running");
+ private:
+  // not thread safe
+  RangeEventList& getEventList(int64_t thread_id = -1) {
+    if (thread_id < 0) {
+      thread_id = RecordFunction::currentThreadId();
+    }
+    auto it = event_lists_map_.find(thread_id);
+    if (it != event_lists_map_.end()) {
+      return *(it->second);
+    } else {
+      auto event_list = std::make_shared<RangeEventList>();
+      event_lists_map_[thread_id] = event_list;
+      return *(event_list.get());
+    }
   }
 
+  std::mutex mutex_;
+  std::unordered_map<uint16_t, std::shared_ptr<RangeEventList>>
+      event_lists_map_;
```

Comment:
I guess I should make it more clearer in the code to avoid confusion.
This struct is shared between main thread and async task, so it's not thread local,
and both main thread and async task write into the same event list inside this struct

re. thread local:
the shared_ptr pointer to the 'current' top of the list is thread local,
in fact on the whole process level we have a tree, each thread has a thread local pointer that points to a node and the thread "sees" nodes on the path from the root to the current node;

Reply:
> This struct is shared between main thread and async task, so it's not thread local,
and both main thread and async task write into the event lists inside this struct

I should've been more clear in my previous comment, yes its true that the structure is shared between multiple threads. I guess my point was that do we need a map indexed by the thread_id? Why can't it just be `std::shared_ptr<RangeEventList>` and each thread that shares the same structure just holds the mutex and writes to the same RangeEventList?

Reply:
it will serialize all the writes in the event list, and the profiler is designed to work in multi-threaded case

Reply:
I see, yes, at this moment the map is redundant because we take a lock anyway, we should either just use one list as you suggested, or switch to a better lock granularity (e.g. per list locks); will address in the next update